### PR TITLE
Support non-deterministic Source 3, by adding a notion of language variants

### PR DESCRIPTION
**Pull request by anubh-v**
#### High level idea:
- Create an array `sourceLanguages` that will hold all the possible languages (similar to the existing `sourceChapters` array that holds the chapter numbers)
- Modify the logic used to generate the chapter dropdown, such that it includes the variants  when creating the dropdown.
- Pass in the variant to the handler function that reacts to the selection of a chapter.
Since some workspaces uses the same handler function, this step requires changes in many workspaces (not just `Playground`, but also `Grading`, `Sourcecast` and `Sourcereel`)

- Create variant as an optional property in the Library type

The variant has to be passed in the library object in two events:

- When the context is "reset" when an option in the chapter dropdown menu is selected
- When the code is evaluated

#### To add a new language:
In `src/reducers/states.ts`, add an entry in the following datstructures:
- `sourceLanguages`: stores languages with their chapter and variant
- `variantDisplay`: stores the user facing names of variants
- `languageURLNames`: stores the names by which languages should be activated, via a url

### CODE REVIEW

**Comment by arsalanc-v2** [comment]
A sample URL for testing language selection through a URL:
```
http://localhost:8075/playground#chap=3_Non_Det&prgrm=GYVwdgxgLglg9mABOGUD6EYCcIBsCmAFFAJSIDeAUIjYlvlCFkgLYCGA1vmgA5wxgohdlAAWaAM4DCAJkQAqRCPEAFAJILEpADTVa%2Bg4drKMcCbM0n1m0iQDclAL6UAJljYB3U2DD5o%2BFzRQXFw0ADcYfA9CAEYABjiSQhR0TBwCeyA
```

File: src/reducers/states.ts [hunk 1]
```diff
@@ -219,7 +219,31 @@ export enum Role {
...
+  { chapter: 3, variant: 'non-det' },
+  { chapter: 4, variant: 'default' }
+];
+
+const variantDisplay: Map<Variant, string> = new Map([['non-det', 'Non-Det']]);
+export const languageURLNames: Map<string, ISourceLanguage> = new Map([
+  ['3_Non_Det', { chapter: 3, variant: 'non-det' }]
+]);
+
+export const styliseChapter = (chap: number, variant?: Variant) => {
```
[the remark this answers is not in the document]
Comment (alcen) [reply src/reducers/states.ts:240]: I think it is a valid suggestion, unless there's some reason that makes it too difficult to enforce?

[the remark this answers is not in the document]
Comment (arsalanc-v2) [reply src/reducers/states.ts]: Yes, using a default variant value for the functions mentioned makes sense.

File: src/sagas/playground.ts [hunk 2]
```diff
@@ -23,13 +25,19 @@ function* updateQueryString() {
...
+  const languageUrlName: string = urlName(chapter, variant);
+
   const external: ExternalLibraryName = yield select(
     (state: IState) => state.workspaces.playground.externalLibrary
   );
   const execTime: number = yield select((state: IState) => state.workspaces.playground.execTime);
   const newQueryString: string = qs.stringify({
     prgrm: compressToEncodedURIComponent(codeString),
-    chap: chapter,
+    chap: languageUrlName,
```
Comment (anubh-v) [inline src/sagas/playground.ts:40]: Should we encode the chapter and variant separately?

Comment (anubh-v) [reply src/sagas/playground.ts:40]: Actually, the current method is simpler for validating the chapter - variant pair in the URL

**Pull request merged**